isabelle/Merkle/MerkleInterpreter.thy

```diff
@@ -0,0 +1,255 @@
+   | None \<Rightarrow> NotReduced)" |
+"reduceMContractStep env state (Pay accId payee token val cont) =
+  (let amountToPay = evalValue env state val in
+   if amountToPay \<le> 0
+   then (let warning = ReduceNonPositivePay accId payee token amountToPay in
+         Reduced warning ReduceNoPayment state cont)
+   else (let balance = moneyInAccount accId token (accounts state);
+             paidAmount = min balance amountToPay;
+             newBalance = balance - paidAmount;
+             newAccs = updateMoneyInAccount accId token newBalance (accounts state);
+             warning = (if paidAmount < amountToPay
+                        then ReducePartialPay accId payee token paidAmount amountToPay
+                        else ReduceNoWarning) ;
+            (payment, finalAccs) = giveMoney accId payee token paidAmount newAccs 
+          in
+         Reduced warning payment (state \<lparr> accounts := finalAccs \<rparr>) cont))" |
+"reduceMContractStep env state (If obs cont1 cont2) =
+  (let cont = (if evalObservation env state obs
+               then cont1
+               else cont2) in
+   Reduced ReduceNoWarning ReduceNoPayment state cont)" |
+"reduceMContractStep env state (When _ timeout cont) =
+  (let (startTime, endTime) = timeInterval env in
+   \<comment> \<open>if timeout in future – do not reduce\<close>
+   if endTime < timeout then NotReduced
+   \<comment> \<open>if timeout in the past – reduce to timeout continuation\<close>   
+   else if timeout \<le> startTime then Reduced ReduceNoWarning ReduceNoPayment state cont
+   \<comment> \<open> if timeout in the time range – issue an ambiguity error\<close>  
+   else AmbiguousTimeIntervalReductionError)" |
+"reduceMContractStep env state (Let valId val cont) =
+  (let evaluatedValue = evalValue env state val;
+       boundVals = boundValues state;
+       newState = state \<lparr> boundValues := MList.insert valId evaluatedValue boundVals \<rparr>;
+       warn = case lookup valId boundVals of
+                Some oldVal \<Rightarrow> ReduceShadowing valId oldVal evaluatedValue
+              | None \<Rightarrow> ReduceNoWarning 
+   in Reduced warn ReduceNoPayment newState cont)" |
+"reduceMContractStep env state (Assert obs cont) =
+  (let warning = if evalObservation env state obs
+                 then ReduceNoWarning
+                 else ReduceAssertionFailed
+   in Reduced warning ReduceNoPayment state cont)"
+
+subsection "Evaluation size"
+
+text \<open>This function gives the notion of size for an evaluation. It is used to show that reductionMLoop
+terminates.\<close>
+fun evalSize :: "State \<Rightarrow> MContract \<Rightarrow> nat" where
+"evalSize sta cont = length (accounts sta) + 2 * (size cont)"
+
+lemma reduceMContractStepReducesSize :
+  "reduceMContractStep env sta c = Reduced twa tef nsta nc \<Longrightarrow>
+     (evalSize nsta nc) < (evalSize sta c)"
+  sorry
```

Comment:
This was copy pasted and adapted from the evalBound in the Semantic module. This should later be modified to use the `mcontract_size` function, currently defined in the Unmerkleize module

Reply:
We shouldn't have any `sorry`s in `main`, we should change it to `oops` and comment out code that depends on it. I guess CI is passing because this is not added to CI yet, otherwise it shouldn't pass

Reply:
I left this as sorry on purpose because it is a draft PR, not intended to be merged on main, rather to start the discussion on the direction we should take to formalize merkleization. The most important thing to assess in this PR is which approach of the ones stated in the Readme should be used (or maybe one that I didn't thought of)

Reply:
On another note, I think the CI tests are only firing when we make a change in a haskell file, we should check this with the help of @shlevy